File: C++/201703/20170304/main.cpp
#include <stdio.h>
#include <limits.h>
#include <algorithm>

#define MAX_N 100001

using namespace std;

struct Edge {
    int ver_adj;
    int cost;
    Edge* link;
    Edge(int v, int c): ver_adj(v), cost(c) {}
};

struct Vertex {
    int ver_name;
    Edge* adjacent;
    Vertex(){}
    Vertex(int n): ver_name(n), adjacent(NULL){}
};

Vertex head[MAX_N];
int n ,m;
int visit[MAX_N];
int dis[MAX_N];

void dijkstra(int v);

int main()
{
    scanf("%d %d", &n, &m);
    fill(visit, visit + MAX_N, 0);      // fill: algorithm
    fill(dis, dis + MAX_N, INT_MAX);    // INT_MAX: limits.h
    for (int i = 1; i <= n; i++)
        head[i] = Vertex(i);
    for (int i = 0; i < m; i++)
    {
        int s, t, c;
        scanf("%d %d %d", &s, &t, &c);
        Edge* e = new Edge(t, c);
        e->link = head[s].adjacent;
        head[s].adjacent = e;

        e = new Edge(s, c);
        e->link = head[t].adjacent;
        head[t].adjacent = e;
    }
    dijkstra(1);
    printf("%d", dis[n]);
    return 0;
}

void dijkstra(int v)
{
    dis[v] = 0;
    for (int i = 0; i < n - 1; i++)
    {
        visit[v] = 1;

        Edge* e = head[v].adjacent;

        while (e != NULL)
        {
            int t = e->ver_adj;
            if (visit[t] == 0 && dis[t] > max(dis[v], e->cost))
                dis[t] = max(dis[v], e->cost);
            e = e->link;
        }

        int u = -1, min_cost = INT_MAX;

        for (int j = 1; j <= n; j++)
            if (visit[j] == 0 && dis[j] < min_cost)
            {
                u = j;
                min_cost = dis[j];
            }
        v = u;
    }
}

File: Java/201703/20170304/Main.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Main {
	// static int[][] graph = new int[100000][100000];
	
	static Vertex[] head = new Vertex[100001];
	static int N, M;
	
	static int[] visited = new int[100001];
	static int[] path = new int[100001];
	static int[] dis = new int[100001];
	
	public static void main(String[] args) throws Exception {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		String line = br.readLine();
		String[] datas = line.split(" ");
		
		N = Integer.parseInt(datas[0]);
		M = Integer.parseInt(datas[1]);
		for (int i = 1; i <= N; i++)
			head[i] = new Vertex(i);
		for (int i = 0; i < M; i++) {
			int a, b, c;
			datas = br.readLine().split(" ");
			a = Integer.parseInt(datas[0]);
			b = Integer.parseInt(datas[1]);
			c = Integer.parseInt(datas[2]);
			Edge edge = new Edge(b, c);
			edge.link = head[a].adjacent;
			head[a].adjacent = edge;
			edge = new Edge(a, c);
			edge.link = head[b].adjacent;
			head[b].adjacent = edge;
		}
		
		dijkstra(1);
		
//		for (int i = 1; i <= N; i++) {
//			System.out.print(path[i] + " ");
//		}
		
		System.out.println(dis[N]);
	}
	
	public static void dijkstra(int v) {
		for (int i = 1; i <= N; i++) {
			visited[i] = 0;
			path[i] = -1;
			dis[i] = Integer.MAX_VALUE;
		}
		
		dis[v] = 0;
		
		for (int i = 0; i < N - 1; i++) {
			visited[v] = 1;
			Vertex vertex = head[v];
			Edge edge = vertex.adjacent;
			while (edge != null) {
				if (visited[edge.verAdj] == 0 &&
					dis[edge.verAdj] > Math.max(edge.cost, dis[v])) {	// 这里的距离应改为路径上最长隧道长度，而不是总长度
					dis[edge.verAdj] = Math.max(edge.cost, dis[v]);
					path[edge.verAdj] = v;
				}
				edge = edge.link;
			}
			
			int minDis = Integer.MAX_VALUE;
			int u = 1;
			
			for (int j = 1; j <= N; j++) {
				if (visited[j] == 0 && minDis > dis[j]) {
					minDis = dis[j];
					u = j;
				}
			}
			
			v = u;
		}
	}
}

class Vertex {
	int verName;
	Edge adjacent;
	public Vertex(int verName) {
		super();
		this.verName = verName;
		adjacent = null;
	}
}

class Edge {
	int verAdj;
	int cost;
	Edge link;
	public Edge(int verAdj, int cost) {
		super();
		this.verAdj = verAdj;
		this.cost = cost;
		link = null;
	}
}